### PR TITLE
CLIPFeatureExtractor should resize images with kept aspect ratio

### LINKED ISSUE #11992: CLIPFeatureExtractor should resize images with kept aspect ratio
## Environment info
- `transformers` version: 4.6.1
- Platform: Linux-5.11.0-7614-generic-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.6.9
- PyTorch version (GPU?): 1.8.1+cu102 (False)
- Tensorflow version (GPU?): 2.5.0 (False)
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help

@patil-suraj 
@LysandreJik 
@patrickvonplaten 
@sgugger 

## Information

Model I am using (Bert, XLNet ...): CLIP

The problem arises when using:
* [x] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [ ] my own task or dataset: (give details below)

## To reproduce

The CLIPFeatureExtractor does not replicate the behavior of the [CLIP reference implementation](https://github.com/openai/CLIP). The below code is taken from the official [huggingface transformers CLIP documentation](https://huggingface.co/transformers/model_doc/clip.html):

```
$ docker run --rm -it huggingface/transformers-cpu:4.6.1
root@02cd404c4a60:/workspace# pip install Pillow==7.2.0
root@02cd404c4a60:/workspace# python3
Python 3.6.9 (default, Jan 26 2021, 15:33:00) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> from PIL import Image
>>> import requests
>>> from transformers import CLIPProcessor, CLIPModel
>>> model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
>>> processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
>>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
>>> image = Image.open(requests.get(url, stream=True).raw)
>>> inputs = processor(images=image, return_tensors="pt", padding=True)
>>> inputs
{'pixel_values': tensor([[[[ 0.2807,  0.3829,  0.4267,  ..., -0.2886, -0.2740, -0.2886],
          [ 0.3245,  0.3829,  0.4121,  ..., -0.2886, -0.2886, -0.3178],
          [ 0.2807,  0.3537,  0.3683,  ..., -0.3762, -0.3470, -0.3178],
          ...,
          [ 1.6384,  1.5362,  1.4194,  ...,  1.3902,  1.2880,  1.2442],
          [ 1.6092,  1.5508,  1.5070,  ...,  1.2150,  0.9814,  0.8501],
          [ 1.6092,  1.4778,  1.4924,  ...,  0.1201, -0.1280, -0.3908]],

         [[-1.3919, -1.3919, -1.3919,  ..., -1.5420, -1.5420, -1.5570],
          [-1.3469, -1.3469, -1.3469,  ..., -1.5270, -1.5120, -1.5270],
          [-1.4069, -1.3769, -1.3469,  ..., -1.5570, -1.5420, -1.5420],
          ...,
          [-0.3414, -0.4614, -0.5515,  ..., -0.6415, -0.7016, -0.7466],
          [-0.3414, -0.3864, -0.4914,  ..., -0.7316, -0.8666, -0.9267],
          [-0.3714, -0.4914, -0.5065,  ..., -1.2869, -1.3769, -1.4820]],

         [[-0.6555, -0.4990, -0.5417,  ..., -1.0110, -0.9256, -0.9541],
          [-0.6981, -0.5986, -0.5701,  ..., -1.0110, -0.9541, -1.0110],
          [-0.6128, -0.5275, -0.4990,  ..., -1.0252, -1.0394, -1.0536],
          ...,
          [ 1.3638,  1.3496,  1.1221,  ...,  1.1647,  1.0652,  0.9514],
          [ 1.3354,  1.1789,  1.2643,  ...,  0.9372,  0.7523,  0.6244],
          [ 1.3780,  1.3780,  1.2643,  ..., -0.1293, -0.5559, -0.7408]]]])}
```

## Expected behavior

Feeding the same image through the official CLIP preprocessing function gives different results:

```
$ docker run --rm -it pytorch/pytorch:1.8.1-cuda11.1-cudnn8-runtime
root@fe597c8c5e9f:/workspace# apt update && apt install git 
root@fe597c8c5e9f:/workspace# pip install ftfy regex tqdm
root@fe597c8c5e9f:/workspace# pip install git+https://github.com/openai/CLIP.git
root@fe597c8c5e9f:/workspace# python
Python 3.8.8 (default, Feb 24 2021, 21:46:12) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> import clip
>>> import requests
>>> from PIL import Image
>>> device = "cpu"
>>> model, preprocess = clip.load("ViT-B/32", device=device)
>>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
>>> inputs = preprocess(Image.open(requests.get(url, stream=True).raw)).unsqueeze(0).to(device)
>>> inputs
tensor([[[[ 0.5873,  0.5873,  0.6165,  ...,  0.0617,  0.0471, -0.0259],
          [ 0.5727,  0.5727,  0.6603,  ...,  0.1201,  0.0763,  0.0909],
          [ 0.5873,  0.5435,  0.6165,  ...,  0.0325,  0.1201,  0.0617],
          ...,
          [ 1.8719,  1.8573,  1.8719,  ...,  1.3902,  1.4340,  1.4194],
          [ 1.8281,  1.8719,  1.8427,  ...,  1.4486,  1.4340,  1.5070],
          [ 1.8573,  1.9011,  1.8281,  ...,  1.3756,  1.3610,  1.4486]],

         [[-1.3169, -1.3019, -1.3169,  ..., -1.4970, -1.4369, -1.4820],
          [-1.2418, -1.2718, -1.2268,  ..., -1.4369, -1.4669, -1.4519],
          [-1.2568, -1.3169, -1.2268,  ..., -1.4669, -1.4069, -1.4519],
          ...,
          [ 0.1239,  0.1089,  0.1239,  ..., -0.7016, -0.6865, -0.6865],
          [ 0.0789,  0.0939,  0.0488,  ..., -0.6565, -0.6865, -0.6115],
          [ 0.0939,  0.1089,  0.0038,  ..., -0.7766, -0.7316, -0.6115]],

         [[-0.4848, -0.4137, -0.3853,  ..., -0.9541, -0.8545, -0.8545],
          [-0.4137, -0.4706, -0.3711,  ..., -0.8119, -0.8545, -0.7834],
          [-0.3284, -0.4422, -0.3853,  ..., -0.8688, -0.8119, -0.8830],
          ...,
          [ 1.5771,  1.6482,  1.6340,  ...,  0.9088,  0.9514,  0.8945],
          [ 1.6198,  1.6055,  1.6055,  ...,  0.8661,  0.8092,  0.7950],
          [ 1.6624,  1.6766,  1.5487,  ...,  0.7950,  0.8661,  0.8519]]]])
```

### PR DESCRIPTION
**Pull request by TobiasNorlund**
# What does this PR do?
Fixes #11992 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@patil-suraj
@LysandreJik
@patrickvonplaten
@sgugger

## Description

With this PR, the preprocessing should match the original preprocessing exactly. With the example from #11992 we now get the exact same results:

```python
>>> import torch
>>> import requests
>>> from transformers import CLIPProcessor, CLIPModel
>>> model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
>>> processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
>>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
>>> image = Image.open(requests.get(url, stream=True).raw)
>>> inputs = processor(images=image, return_tensors="pt", padding=True)
>>> inputs
{'pixel_values': tensor([[[[ 0.5873,  0.5873,  0.6165,  ...,  0.0617,  0.0471, -0.0259],
          [ 0.5727,  0.5727,  0.6603,  ...,  0.1201,  0.0763,  0.0909],
          [ 0.5873,  0.5435,  0.6165,  ...,  0.0325,  0.1201,  0.0617],
          ...,
          [ 1.8719,  1.8573,  1.8719,  ...,  1.3902,  1.4340,  1.4194],
          [ 1.8281,  1.8719,  1.8427,  ...,  1.4486,  1.4340,  1.5070],
          [ 1.8573,  1.9011,  1.8281,  ...,  1.3756,  1.3610,  1.4486]],

         [[-1.3169, -1.3019, -1.3169,  ..., -1.4970, -1.4369, -1.4820],
          [-1.2418, -1.2718, -1.2268,  ..., -1.4369, -1.4669, -1.4519],
          [-1.2568, -1.3169, -1.2268,  ..., -1.4669, -1.4069, -1.4519],
          ...,
          [ 0.1239,  0.1089,  0.1239,  ..., -0.7016, -0.6865, -0.6865],
          [ 0.0789,  0.0939,  0.0488,  ..., -0.6565, -0.6865, -0.6115],
          [ 0.0939,  0.1089,  0.0038,  ..., -0.7766, -0.7316, -0.6115]],

         [[-0.4848, -0.4137, -0.3853,  ..., -0.9541, -0.8545, -0.8545],
          [-0.4137, -0.4706, -0.3711,  ..., -0.8119, -0.8545, -0.7834],
          [-0.3284, -0.4422, -0.3853,  ..., -0.8688, -0.8119, -0.8830],
          ...,
          [ 1.5771,  1.6482,  1.6340,  ...,  0.9088,  0.9514,  0.8945],
          [ 1.6198,  1.6055,  1.6055,  ...,  0.8661,  0.8092,  0.7950],
          [ 1.6624,  1.6766,  1.5487,  ...,  0.7950,  0.8661,  0.8519]]]])}
```

### CODE REVIEW

**Comment by patil-suraj** [comment]
Hi @TobiasNorlund , thanks a lot for spotting this.

However, this change does not actually work as expected. I processed a few images using the origin CLIP transforms and `CLIPFeatureExtractor` and compared the output. Here's the script

```python3
from PIL import Image
import os
import skimage

import torch
from clip import load

from transformers import CLIPConfig, CLIPModel, CLIPTokenizer, CLIPFeatureExtractor, CLIPProcessor

_, transforms = load("./model.pt", jit=False)

proc = CLIPProcessor.from_pretrained("./clip-vit-base-patch32/")

files = [filename for filename in os.listdir(skimage.data_dir) if filename.endswith(".png") or filename.endswith(".jpg")]

images = []
for filename in files:
    image = transforms(Image.open(os.path.join(skimage.data_dir, filename)).convert("RGB"))
    images.append(image)

hf_images = []
for filename in files:
    image = Image.open(os.path.join(skimage.data_dir, filename)).convert("RGB")
    enc = proc(images=image, return_tensors="pt")
    hf_images.append(enc.pixel_values.squeeze(0))

match = [torch.allclose(hf_image, pt_image, atol=4e-2) for hf_image, pt_image in zip(hf_images, images)]
all(match)
```

I looked into `torchvision' s` `resize` and `center_crop` implementation and turns out they are a bit different than the way we have implemented it.  Following their implem I tried overriding the `resize` and `center_crop` methods in `CLIPFeatureExtractor` which seems to be working. Here's the code 
 
```python
def center_crop(self, image, size):
    """
    Crops :obj:`image` to the given size using a center crop. Note that if the image is too small to be cropped to
    the size is given, it will be padded (so the returned result has the size asked).

    Args:
        image (:obj:`PIL.Image.Image` or :obj:`np.ndarray` or :obj:`torch.Tensor`):
            The image to resize.
        size (:obj:`int` or :obj:`Tuple[int, int]`):
            The size to which crop the image.
    """
    self._ensure_format_supported(image)
    if not isinstance(size, tuple):
        size = (size, size)

    if not isinstance(image, Image.Image):
        image = self.to_pil_image(image)

    image_width, image_height = image.size
    crop_height, crop_width = size

    crop_top = int((image_height - crop_height + 1) * 0.5)
    crop_left = int((image_width - crop_width + 1) * 0.5)

    return image.crop((crop_left, crop_top, crop_left + crop_width, crop_top + crop_height))

def resize(self, image, size, resample=Image.BICUBIC):
    width, height = image.size

    short, long = (width, height) if width <= height else (height, width)
    if short == size:
        return image

    new_short, new_long = size, int(size * long / short)

    new_w, new_h = (new_short, new_long) if width <= height else (new_long, new_short)
    return image.resize((new_w, new_h), resample)
```
With this change, there is no need to change the `__call__` method.
However, this change requires `center_crop`  to be always applied as `resize` won't always resize to an exact given size. 


Could you verify this on your end and update the PR? Thanks!

**Comment by TobiasNorlund** [comment]
Thanks @patil-suraj !
Please have a look at the updated PR, in which I modified your `resize` method to also support non-PIL images to make tests pass.

File: tests/test_modeling_clip.py [hunk 1]
```diff
@@ -556,6 +557,6 @@ def test_inference(self):
             torch.Size((inputs.input_ids.shape[0], inputs.pixel_values.shape[0])),
         )
 
-        expected_logits = torch.tensor([[24.5056, 18.8076]], device=torch_device)
+        expected_logits = torch.Tensor([[24.5701, 19.3049]]).to(torch_device)
```
Comment (patil-suraj) [inline tests/test_modeling_clip.py:560]: The new pre-preprocessing slightly changes the `logits`

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for fixing this @TobiasNorlund !

File: tests/test_modeling_clip.py (hunk 1 above)
Comment (sgugger) [reply tests/test_modeling_clip.py:560]: We shouldn't use torch.Tensor anymore, see #12027

**Review by sgugger — APPROVED** [review]
Thanks for fixing this!

File: tests/test_modeling_clip.py (hunk 1 above)
Comment (patil-suraj) [inline tests/test_modeling_clip.py:560]:
```suggestion
        expected_logits = torch.tensor([[24.5701, 19.3049]], device=torch_device)
```
